seml/start.py

```diff
@@ -338,124 +341,218 @@ def batch_chunks(exp_chunks):
-                                               unobserved=unobserved, post_mortem=post_mortem)
-            if 'conda_environment' in exp['seml']:
-                configs.append((exe, exp['seml']['conda_environment'], config))
-            else:
-                configs.append((exe, None, config))
-        return configs
-    elif slurm:
-        if not output_to_file:
-            logging.error("Output cannot be written to stdout in Slurm mode. "
-                          "Remove the '--output-to-console' argument.")
-            sys.exit(1)
-        exp_chunks = chunk_list(exps_list)
-        exp_arrays = batch_chunks(exp_chunks)
-        njobs = len(exp_chunks)
-        narrays = len(exp_arrays)
-
-        logging.info(f"Starting {nexps} experiment{s_if(nexps)} in "
-                     f"{njobs} Slurm job{s_if(njobs)} in {narrays} Slurm job array{s_if(narrays)}.")
-
-        for exp_array in exp_arrays:
-            job_name = get_exp_name(exp_array[0][0], collection.name)
-            output_dir_path = get_output_dir_path(exp_array[0][0])
-            slurm_config = exp_array[0][0]['slurm']
-            del slurm_config['experiments_per_job']
-            start_slurm_job(collection, exp_array, unobserved, post_mortem,
-                            name=job_name, output_dir_path=output_dir_path, **slurm_config)
+    num_exceptions = 0
+    jobs_counter = 0
+
+    if not local_steal_slurm_jobs:
+        staged_query = {'status': {"$in": States.PENDING}, 'slurm.array_id': {'$exists': False}}
     else:
-        login_node_name = 'fs'
-        if login_node_name in os.uname()[1]:
-            logging.error("Refusing to run a compute experiment on a login node. "
-                          "Please use Slurm or a compute node.")
-            sys.exit(1)
-        # [get_output_dir_path(exp) for exp in exps_list]  # Check if output dir exists
-        logging.info(f'Starting local worker thread that will run up to {nexps} experiment{s_if(nexps)}, '
-                     f'until no staged experiments remain.')
-        if not unobserved:
-            collection.update_many({'_id': {'$in': [e['_id'] for e in exps_list if e['status'] in States.STAGED]}},
-                                   {"$set": {"status": States.PENDING[0]}})
-        num_exceptions = 0
-        tq = tqdm(enumerate(exps_list))
-        for i_exp, exp in tq:
-            if output_to_file:
-                output_dir_path = get_output_dir_path(exp)
-            else:
-                output_dir_path = None
-            success = start_local_job(collection, exp, unobserved, post_mortem, output_dir_path)
+        staged_query = {'status': {"$in": States.PENDING}}
+
+    staged_query.update(filter_dict)
+
+    tq = tqdm()
+    while collection.count_documents(staged_query) > 0 and jobs_counter < num_exps:
+        exp = collection.find_one_and_update(staged_query, {"$set": {"status": States.RUNNING[0]}})
+        if exp['status'] not in States.PENDING:
```

Comment:
We don't need this, we've already checked for the state.

Reply:
Ok, I'm replacing line 510 with `if exp is None: continue`.

Reply:
Fixed in 2970ece.